# Data Exploration and Cleaning

In this notebook, we load and understand the data, and clean and export it to JSON format, to be used with the APIs.

As a pre-requisite, 
- Install `jupyter`
- Install `pandas` via `pip` (`numpy` will be installed automatically as a dependency)
- Make sure that the `data` folder contains the dataset

## Data Loading

In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
import json

In [4]:
data = pd.read_csv("data/greenhouse_gas_inventory_data_data.csv")

In [5]:
data

country_or_area  year          value  \
0                    Australia  2014  393126.946994   
1                    Australia  2013  396913.936530   
2                    Australia  2012  406462.847704   
3                    Australia  2011  403705.528314   
4                    Australia  2010  406200.993184   
...                        ...   ...            ...   
8401  United States of America  1994     593.222570   
8402  United States of America  1993     586.939752   
8403  United States of America  1992     574.414624   
8404  United States of America  1991     566.900968   
8405  United States of America  1990     293.205663   

                                               category  
0     carbon_dioxide_co2_emissions_without_land_use_...  
1     carbon_dioxide_co2_emissions_without_land_use_...  
2     carbon_dioxide_co2_emissions_without_land_use_...  
3     carbon_dioxide_co2_emissions_without_land_use_...  
4     carbon_dioxide_co2_emissions_without_land_use_...  
...                                                 ...  
8401  unspecified_mix_of_hydrofluorocarbons_hfcs_and...  
8402  unspecified_mix_of_hydrofluorocarbons_hfcs_and...  
8403  unspecified_mix_of_hydrofluorocarbons_hfcs_and...  
8404  unspecified_mix_of_hydrofluorocarbons_hfcs_and...  
8405  unspecified_mix_of_hydrofluorocarbons_hfcs_and...  

[8406 rows x 4 columns]

We see that the dataset contains the country name, along with emission values belonging to several categories.

Let's dig deep into see which categories are present

## Unique Categories

In [6]:
data['category'].unique()

array(['carbon_dioxide_co2_emissions_without_land_use_land_use_change_and_forestry_lulucf_in_kilotonne_co2_equivalent',
       'greenhouse_gas_ghgs_emissions_including_indirect_co2_without_lulucf_in_kilotonne_co2_equivalent',
       'greenhouse_gas_ghgs_emissions_without_land_use_land_use_change_and_forestry_lulucf_in_kilotonne_co2_equivalent',
       'hydrofluorocarbons_hfcs_emissions_in_kilotonne_co2_equivalent',
       'methane_ch4_emissions_without_land_use_land_use_change_and_forestry_lulucf_in_kilotonne_co2_equivalent',
       'nitrogen_trifluoride_nf3_emissions_in_kilotonne_co2_equivalent',
       'nitrous_oxide_n2o_emissions_without_land_use_land_use_change_and_forestry_lulucf_in_kilotonne_co2_equivalent',
       'perfluorocarbons_pfcs_emissions_in_kilotonne_co2_equivalent',
       'sulphur_hexafluoride_sf6_emissions_in_kilotonne_co2_equivalent',
       'unspecified_mix_of_hydrofluorocarbons_hfcs_and_perfluorocarbons_pfcs_emissions_in_kilotonne_co2_equivalent'],
      dtype=obj

We see some of the categories listing chemicals or chemical groups, while others are generals. We shall pick following chemicals and chemical groups for this exercise. We shall also rename them afterwards.

- CO<sub>2</sub>
- HFCs
- CH<sub>4</sub>
- NF<sub>3</sub>
- N<sub>2</sub>O
- PFCs
- SF<sub>6</sub>

Since all of the of the units are in _kilotonnes of CO2 equivalent_, we can use them directly, without any modification.

## Picking and Renaming the Categories

We use the `new_data` variable for the cleaned dataset. It will be re-assigned several times in the future usages.

In [7]:
new_data = data.loc[data['category'].isin([
    'carbon_dioxide_co2_emissions_without_land_use_land_use_change_and_forestry_lulucf_in_kilotonne_co2_equivalent',
    'hydrofluorocarbons_hfcs_emissions_in_kilotonne_co2_equivalent',
    'methane_ch4_emissions_without_land_use_land_use_change_and_forestry_lulucf_in_kilotonne_co2_equivalent',
    'nitrogen_trifluoride_nf3_emissions_in_kilotonne_co2_equivalent',
    'nitrous_oxide_n2o_emissions_without_land_use_land_use_change_and_forestry_lulucf_in_kilotonne_co2_equivalent',
    'perfluorocarbons_pfcs_emissions_in_kilotonne_co2_equivalent',
    'sulphur_hexafluoride_sf6_emissions_in_kilotonne_co2_equivalent',
])]
new_data

country_or_area  year          value  \
0                    Australia  2014  393126.946994   
1                    Australia  2013  396913.936530   
2                    Australia  2012  406462.847704   
3                    Australia  2011  403705.528314   
4                    Australia  2010  406200.993184   
...                        ...   ...            ...   
8326  United States of America  1994   28004.247351   
8327  United States of America  1993   29432.652199   
8328  United States of America  1992   29949.821295   
8329  United States of America  1991   29764.563408   
8330  United States of America  1990   31080.406159   

                                               category  
0     carbon_dioxide_co2_emissions_without_land_use_...  
1     carbon_dioxide_co2_emissions_without_land_use_...  
2     carbon_dioxide_co2_emissions_without_land_use_...  
3     carbon_dioxide_co2_emissions_without_land_use_...  
4     carbon_dioxide_co2_emissions_without_land_use_...  
...                                                 ...  
8326  sulphur_hexafluoride_sf6_emissions_in_kilotonn...  
8327  sulphur_hexafluoride_sf6_emissions_in_kilotonn...  
8328  sulphur_hexafluoride_sf6_emissions_in_kilotonn...  
8329  sulphur_hexafluoride_sf6_emissions_in_kilotonn...  
8330  sulphur_hexafluoride_sf6_emissions_in_kilotonn...  

[6308 rows x 4 columns]

In [8]:
new_data = new_data.replace({
    'category': {
        'carbon_dioxide_co2_emissions_without_land_use_land_use_change_and_forestry_lulucf_in_kilotonne_co2_equivalent': 'CO2',
        'hydrofluorocarbons_hfcs_emissions_in_kilotonne_co2_equivalent': 'HFCs',
        'methane_ch4_emissions_without_land_use_land_use_change_and_forestry_lulucf_in_kilotonne_co2_equivalent': 'CH4',
        'nitrogen_trifluoride_nf3_emissions_in_kilotonne_co2_equivalent': 'NF3',
        'nitrous_oxide_n2o_emissions_without_land_use_land_use_change_and_forestry_lulucf_in_kilotonne_co2_equivalent': 'N2O',
        'perfluorocarbons_pfcs_emissions_in_kilotonne_co2_equivalent': 'PFCs',
        'sulphur_hexafluoride_sf6_emissions_in_kilotonne_co2_equivalent': 'SF6',
    }
})
new_data

country_or_area  year          value category
0                    Australia  2014  393126.946994      CO2
1                    Australia  2013  396913.936530      CO2
2                    Australia  2012  406462.847704      CO2
3                    Australia  2011  403705.528314      CO2
4                    Australia  2010  406200.993184      CO2
...                        ...   ...            ...      ...
8326  United States of America  1994   28004.247351      SF6
8327  United States of America  1993   29432.652199      SF6
8328  United States of America  1992   29949.821295      SF6
8329  United States of America  1991   29764.563408      SF6
8330  United States of America  1990   31080.406159      SF6

[6308 rows x 4 columns]

In [9]:
new_data['category'].unique()

array(['CO2', 'HFCs', 'CH4', 'NF3', 'N2O', 'PFCs', 'SF6'], dtype=object)

Good! Now that the catgories are looking clean, we shall look at the countries.

In [10]:
new_data['country_or_area'].unique()

array(['Australia', 'Austria', 'Belarus', 'Belgium', 'Bulgaria', 'Canada',
       'Croatia', 'Cyprus', 'Czech Republic', 'Denmark', 'Estonia',
       'European Union', 'Finland', 'France', 'Germany', 'Greece',
       'Hungary', 'Iceland', 'Ireland', 'Italy', 'Japan', 'Latvia',
       'Liechtenstein', 'Lithuania', 'Luxembourg', 'Malta', 'Monaco',
       'Netherlands', 'New Zealand', 'Norway', 'Poland', 'Portugal',
       'Romania', 'Russian Federation', 'Slovakia', 'Slovenia', 'Spain',
       'Sweden', 'Switzerland', 'Turkey', 'Ukraine', 'United Kingdom',
       'United States of America'], dtype=object)

Here, we observe some things:
- Countries are represented by the name
- Eurpoean Union data is present, along with its member countries' (See Gernmany, Italy, Netherlands etc.)

## Cleaning the countries
To make the data consistent,
- Remove EU from the data
- Convert country names to ISO A3 3-letter country code format. (We do this, as the [GeoJSON of country borders we would be using in the frontend](https://datahub.io/core/geo-countries) uses ISO A3 country codes)


In [11]:
new_data = new_data.loc[data['country_or_area']!= "European Union"]
new_data

country_or_area  year          value category
0                    Australia  2014  393126.946994      CO2
1                    Australia  2013  396913.936530      CO2
2                    Australia  2012  406462.847704      CO2
3                    Australia  2011  403705.528314      CO2
4                    Australia  2010  406200.993184      CO2
...                        ...   ...            ...      ...
8326  United States of America  1994   28004.247351      SF6
8327  United States of America  1993   29432.652199      SF6
8328  United States of America  1992   29949.821295      SF6
8329  United States of America  1991   29764.563408      SF6
8330  United States of America  1990   31080.406159      SF6

[6133 rows x 4 columns]

As a verification, we look at the countries again

In [12]:
new_data['country_or_area'].unique()

array(['Australia', 'Austria', 'Belarus', 'Belgium', 'Bulgaria', 'Canada',
       'Croatia', 'Cyprus', 'Czech Republic', 'Denmark', 'Estonia',
       'Finland', 'France', 'Germany', 'Greece', 'Hungary', 'Iceland',
       'Ireland', 'Italy', 'Japan', 'Latvia', 'Liechtenstein',
       'Lithuania', 'Luxembourg', 'Malta', 'Monaco', 'Netherlands',
       'New Zealand', 'Norway', 'Poland', 'Portugal', 'Romania',
       'Russian Federation', 'Slovakia', 'Slovenia', 'Spain', 'Sweden',
       'Switzerland', 'Turkey', 'Ukraine', 'United Kingdom',
       'United States of America'], dtype=object)

Now that the the list contain all the countries, we shall convert them into A3 codes using [this list](https://www.rocketroute.com/iso-country-code-list). We may not need to retain the name, as the GeoJSON already contains it.

In [13]:
country_map = {
       'Australia': 'AUS', 'Austria': 'AUT', 'Belarus': 'BLR', 'Belgium': 'BEL', 'Bulgaria': 'BGR', 'Canada': 'CAN',
       'Croatia': 'HRV', 'Cyprus': 'CYP', 'Czech Republic': 'CZE', 'Denmark': 'DNK', 'Estonia': 'EST',
       'Finland': 'FIN', 'France': 'FRA', 'Germany': 'DEU', 'Greece': 'GRC',
       'Hungary': 'HUN', 'Iceland': 'ISL', 'Ireland': 'IRL', 'Italy': 'ITA', 'Japan': 'JPN', 'Latvia': 'LVA',
       'Liechtenstein': 'LIE', 'Lithuania': 'LTU', 'Luxembourg': 'LUX', 'Malta': 'MLT', 'Monaco': 'MCO',
       'Netherlands': 'NLD', 'New Zealand': 'NZL', 'Norway': 'NOR', 'Poland': 'POL', 'Portugal': 'PRT',
       'Romania': 'ROU', 'Russian Federation': 'RUS', 'Slovakia': 'SVK', 'Slovenia':'SVN', 'Spain': 'ESP',
       'Sweden': 'SWE', 'Switzerland': 'CHE', 'Turkey': 'TUR', 'Ukraine': 'UKR', 'United Kingdom': 'GBR',
       'United States of America': 'USA'
}

In [14]:
new_data['country_code'] = new_data['country_or_area'].map(country_map)
new_data

/tmp/ipykernel_12930/501431545.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_data['country_code'] = new_data['country_or_area'].map(country_map)


country_or_area  year          value category country_code
0                    Australia  2014  393126.946994      CO2          AUS
1                    Australia  2013  396913.936530      CO2          AUS
2                    Australia  2012  406462.847704      CO2          AUS
3                    Australia  2011  403705.528314      CO2          AUS
4                    Australia  2010  406200.993184      CO2          AUS
...                        ...   ...            ...      ...          ...
8326  United States of America  1994   28004.247351      SF6          USA
8327  United States of America  1993   29432.652199      SF6          USA
8328  United States of America  1992   29949.821295      SF6          USA
8329  United States of America  1991   29764.563408      SF6          USA
8330  United States of America  1990   31080.406159      SF6          USA

[6133 rows x 5 columns]

In [15]:
new_data['country_code'].unique()

array(['AUS', 'AUT', 'BLR', 'BEL', 'BGR', 'CAN', 'HRV', 'CYP', 'CZE',
       'DNK', 'EST', 'FIN', 'FRA', 'DEU', 'GRC', 'HUN', 'ISL', 'IRL',
       'ITA', 'JPN', 'LVA', 'LIE', 'LTU', 'LUX', 'MLT', 'MCO', 'NLD',
       'NZL', 'NOR', 'POL', 'PRT', 'ROU', 'RUS', 'SVK', 'SVN', 'ESP',
       'SWE', 'CHE', 'TUR', 'UKR', 'GBR', 'USA'], dtype=object)

Now, we shall look at the `value` column, which is the emission value in Kilotonnes of CO<sub>2</sub> equivalent.

The value `29764.563408` kilotonne has 6 significant digits after decimal point, meaning it is precise upto a kg. We may not need that precision, so we keep 2 significant difits after decimal point, meaning, it is precise upto 10 tons.

## Rounding the emission `value`

In [16]:
new_data['value'] = np.round(new_data['value'],2)
new_data

/tmp/ipykernel_12930/3892214510.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_data['value'] = np.round(new_data['value'],2)


country_or_area  year      value category country_code
0                    Australia  2014  393126.95      CO2          AUS
1                    Australia  2013  396913.94      CO2          AUS
2                    Australia  2012  406462.85      CO2          AUS
3                    Australia  2011  403705.53      CO2          AUS
4                    Australia  2010  406200.99      CO2          AUS
...                        ...   ...        ...      ...          ...
8326  United States of America  1994   28004.25      SF6          USA
8327  United States of America  1993   29432.65      SF6          USA
8328  United States of America  1992   29949.82      SF6          USA
8329  United States of America  1991   29764.56      SF6          USA
8330  United States of America  1990   31080.41      SF6          USA

[6133 rows x 5 columns]

Now that the data is looking good, we shall export it to a json, which the `nodejs` program will use to initialize the database.

## Database Export

Before that, we have to think about the schema a little bit. Since the goal of the project is to provide 2 APIs
- `GET /countries`
- `GET /countries/{id}?query` (I have taken the freedom to change `/country/{id}` to `countries/{id}` as they refer to the same resource at different levels)

We shall provide two tables
1. `countries` -- `country_id` as the sole-column
2. `emissions` -- `country_id` and `year` as primary key, `emission_values: {[chemical]: number}` as a JSON value (`sqlite` does not support them, so `string` and `JSON.parse` in the code).

We choose JSON for emissions values because we may add some extra key in the future, making it easy to extend the data, without modifying the database

We will write these models as a part of NodeJS source code, and create appropriate views of them to support the use-case.

In [17]:
countris = [{'id': y, 'name': x} for x,y in country_map.items()]
countris[:10]

[{'id': 'AUS', 'name': 'Australia'},
 {'id': 'AUT', 'name': 'Austria'},
 {'id': 'BLR', 'name': 'Belarus'},
 {'id': 'BEL', 'name': 'Belgium'},
 {'id': 'BGR', 'name': 'Bulgaria'},
 {'id': 'CAN', 'name': 'Canada'},
 {'id': 'HRV', 'name': 'Croatia'},
 {'id': 'CYP', 'name': 'Cyprus'},
 {'id': 'CZE', 'name': 'Czech Republic'},
 {'id': 'DNK', 'name': 'Denmark'}]

In [18]:
new_data

country_or_area  year      value category country_code
0                    Australia  2014  393126.95      CO2          AUS
1                    Australia  2013  396913.94      CO2          AUS
2                    Australia  2012  406462.85      CO2          AUS
3                    Australia  2011  403705.53      CO2          AUS
4                    Australia  2010  406200.99      CO2          AUS
...                        ...   ...        ...      ...          ...
8326  United States of America  1994   28004.25      SF6          USA
8327  United States of America  1993   29432.65      SF6          USA
8328  United States of America  1992   29949.82      SF6          USA
8329  United States of America  1991   29764.56      SF6          USA
8330  United States of America  1990   31080.41      SF6          USA

[6133 rows x 5 columns]

In [19]:
new_data["emissions"] = new_data.apply(lambda row: [row["category"], row["value"]], axis=1)
new_data

/tmp/ipykernel_12930/2864945010.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_data["emissions"] = new_data.apply(lambda row: [row["category"], row["value"]], axis=1)


country_or_area  year      value category country_code  \
0                    Australia  2014  393126.95      CO2          AUS   
1                    Australia  2013  396913.94      CO2          AUS   
2                    Australia  2012  406462.85      CO2          AUS   
3                    Australia  2011  403705.53      CO2          AUS   
4                    Australia  2010  406200.99      CO2          AUS   
...                        ...   ...        ...      ...          ...   
8326  United States of America  1994   28004.25      SF6          USA   
8327  United States of America  1993   29432.65      SF6          USA   
8328  United States of America  1992   29949.82      SF6          USA   
8329  United States of America  1991   29764.56      SF6          USA   
8330  United States of America  1990   31080.41      SF6          USA   

             emissions  
0     [CO2, 393126.95]  
1     [CO2, 396913.94]  
2     [CO2, 406462.85]  
3     [CO2, 403705.53]  
4     [CO2, 406200.99]  
...                ...  
8326   [SF6, 28004.25]  
8327   [SF6, 29432.65]  
8328   [SF6, 29949.82]  
8329   [SF6, 29764.56]  
8330   [SF6, 31080.41]  

[6133 rows x 6 columns]

In [20]:
emissions_data = new_data.groupby(["country_code", "year"])["emissions"].apply(list)
emissions_data

country_code  year
AUS           1990    [[CO2, 278265.9], [HFCs, 1424.68], [CH4, 11876...
              1991    [[CO2, 279741.64], [HFCs, 1424.68], [CH4, 1177...
              1992    [[CO2, 284766.09], [HFCs, 1333.18], [CH4, 1168...
              1993    [[CO2, 289142.27], [HFCs, 1829.87], [CH4, 1138...
              1994    [[CO2, 293830.71], [HFCs, 1027.34], [CH4, 1109...
                                            ...                        
USA           2010    [[CO2, 5688756.01], [HFCs, 141633.8], [CH4, 72...
              2011    [[CO2, 5559507.66], [HFCs, 146080.6], [CH4, 71...
              2012    [[CO2, 5349220.95], [HFCs, 147249.75], [CH4, 7...
              2013    [[CO2, 5502550.71], [HFCs, 149922.39], [CH4, 7...
              2014    [[CO2, 5556006.58], [HFCs, 157237.34], [CH4, 7...
Name: emissions, Length: 1049, dtype: object

In [21]:
emissions = [ {"country_id": key[0], "year": key[1], "emissions": dict(value)} for key, value in emissions_data.to_dict().items()]
emissions

[{'country_id': 'AUS',
  'year': 1990,
  'emissions': {'CO2': 278265.9,
   'HFCs': 1424.68,
   'CH4': 118768.84,
   'N2O': 15345.6,
   'PFCs': 4607.01,
   'SF6': 211.02}},
 {'country_id': 'AUS',
  'year': 1991,
  'emissions': {'CO2': 279741.64,
   'HFCs': 1424.68,
   'CH4': 117744.64,
   'N2O': 14923.41,
   'PFCs': 4610.74,
   'SF6': 228.94}},
 {'country_id': 'AUS',
  'year': 1992,
  'emissions': {'CO2': 284766.09,
   'HFCs': 1333.18,
   'CH4': 116837.51,
   'N2O': 15293.18,
   'PFCs': 4603.28,
   'SF6': 246.86}},
 {'country_id': 'AUS',
  'year': 1993,
  'emissions': {'CO2': 289142.27,
   'HFCs': 1829.87,
   'CH4': 113837.51,
   'N2O': 15374.99,
   'PFCs': 3315.53,
   'SF6': 264.76}},
 {'country_id': 'AUS',
  'year': 1994,
  'emissions': {'CO2': 293830.71,
   'HFCs': 1027.34,
   'CH4': 110997.41,
   'N2O': 15790.43,
   'PFCs': 2164.43,
   'SF6': 282.64}},
 {'country_id': 'AUS',
  'year': 1995,
  'emissions': {'CO2': 305162.54,
   'HFCs': 1004.03,
   'CH4': 109900.74,
   'N2O': 15578.2,

We have the data in required format now, let's finish off by writing it to JSON files.

In [22]:
json.dump(countris, open("data/countries.json", "w"), indent=2)
json.dump(emissions, open("data/emissions.json", "w"), indent=2)

This marks the end of our data cleaning journey. We shall define Models, Views and Controllers in the NodeJS code. After that, we shall also write a simple frontend to explore the data.